### PR TITLE
KB-12053: Automate the generation and validation of API docs

### PR DESCRIPTION
**Pull request by carla-at-wiris**
## Automate the generation and validation of the documentation static site generated by the jsdoc library

The `mathtype-html-integration-devkit` package uses the jsdoc library to generate an HTML static site with the API documentation build from the source code comments.

````bash
packages/mathtype-html-integration-devkit$ npm run build-jsdoc
````

> The `build-jsdoc` command generates an static site on the `packages/mathtype-html-integration-devkit/out` folder

This HTML static site will be available to the audience on the next public address: [https://docs.wiris.com/api/integrations/jsdoc/](https://docs.wiris.com/api/integrations/jsdoc/)

> To publish this site to the docs.wiris.com is out of the scope of this task.

This PR adds a new command to the main project that validates the jsdoc generated site using the [link-checker](https://www.npmjs.com/package/link-checker) library.

````bash
$ npm run build-jsdoc-validation
````

> [link-checker](https://www.npmjs.com/package/link-checker) is a well maintaned library that adds no vulnerability risks to the project.

This PR also adds a new GitHub Actions workflow called `generate-docs`.

This action uses the above mentioned commands to generate and validate the jsdoc generated HTML site with the API documentation of the [mathtype-html-integration-devkit](https://www.npmjs.com/package/@wiris/mathtype-html-integration-devkit) package.

The site is then zipped and published as an artifact so it can be consumed by other workflows or manually downloaded.

#### Fix lint errors on the Cypress test implementation

  We've detected some unfixed lint errors from the Cypress test suite implemetation.

  Even though it's not part of this task, we've decided to fix them, since it's not allowed to introduce lint errors to the code base.

  Therefore, it becomes expedite to fix them.

  We've validated that the whole suite of tests work fine and that all tests pass.

#### Fix lint errors on the TinyMCE package

  Fix lint errors on this published package. This fix was lost from one the merged branches before releasing.

#### Apply jsdoc lint validations to devkit only

  Improve the linting strategy to avoid using the jsdoc linting validations to other files than the MathType devkit package.

  Since it's the only package that we generate the jsdoc API static site from.

  Therefore, it makes no sense to lint, using the jsdoc validations, all the JavaScript files in the project that we don't use to generate a jsdoc generated API documentation site.

## More information

The workflow, executed through GitHub actions, is triggered on:
* Every push
* Can be triggered manually
* Can be triggered by other workflows

---

[#taskid 12053](https://wiris.kanbanize.com/ctrl_board/2/cards/12053/details/)

### CODE REVIEW

**Review by xripoll-at-wiris — APPROVED** [review]

**Comment by dcanet-at-wiris** [comment]
Hi!

I've added a lot of changes to this branch.

I've also rebased the commit log to avoid adding unnecessary messages to the log history.

Therefore, I'm adding some comments here.

**Comment by dcanet-at-wiris** [comment]
First, I'm going to discuss what was included on this PR and has been updated:

### Rename commands to a more semantic versions.

I've changed some commands names on the package.json and the github actions workflow to a more sematnci versions.
Also I took the change to improve the comments on the workflow to make it more understandable and follow the same standards as the already created workflows in the project.


### Fix lint errors on the Cypress test implementation

I've detected some unfixed lint errors from the Cypress test suite implementation.

Even though it's not part of this task, we've decided to fix them, since it's not allowed to introduce lint errors to the code base.
Therefore, it becomes expedite to fix them.
We've validated that the whole suite of tests work fine and that all tests pass.

### Fix lint errors on the TinyMCE package

Fix lint errors on this published package. This fix was lost from one the merged branches before releasing.

### Apply jsdoc lint validations to devkit only

Improve the linting strategy to avoid using the jsdoc linting validations to other files than the MathType devkit package.

Since it's the only package that we generate the jsdoc API static site from.

Therefore, it makes no sense to lint, using the jsdoc validations all the JavaScript files in the project that we don't use to generate a jsdoc generated API documentation site.

**Review by dcanet-at-wiris — APPROVED** [review]

File: .github/workflows/jsdoc-Generate-documentation-site.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+      # 05. If the main index.html file is empty, raise an error to stop
+      #     the workflow execution.
+      - name: Errors if jsdoc site is empty
+        if: steps.check_files.outputs.files_exists == 'false'
+        # Only runs if the file does not exist
+        uses: actions/github-script@v3
+        with:
+          script: |
+              core.setFailed('jsdoc: main index file is empty')
```
Comment (carla-at-wiris) [inline .github/workflows/jsdoc-Generate-documentation-site.yml:64]: I believe that the element passed on this function has to be the name of the step you want to force fail. Have you checked what happens when there's no index.html with the change you added in this line?

File: package.json [hunk 2]
```diff
@@ -32,10 +34,12 @@
     "eslint-config-airbnb-base": "^14.2.1",
     "eslint-plugin-cypress": "^2.11.3",
     "eslint-plugin-import": "^2.22.1",
+    "eslint-plugin-jsdoc": "^36.0.6",
     "fs-extra": "^10.0.0",
-    "html-validate": "^5.2.0",
```
Comment (carla-at-wiris) [inline package.json:36]: Is there any reason why you downgraded the html-validate version? This upgrade was made through the dependabot and checked it worked properly.

Comment (dcanet-at-wiris) [reply package.json:36]: Truth is I didn't paid attention to it.
I'm bumping the version to the latest.

**Comment by xripoll-at-wiris** [comment]
I guess this is expected, but the `npm run lint` command in the root fails with the `packages/mathtype-html-integration-devkit/out/styles/main.css` file, after generating the docs. Should this be ignored?

**Comment by dcanet-at-wiris** [comment]
> I guess this is expected, but the `npm run lint` command in the root fails with the `packages/mathtype-html-integration-devkit/out/styles/main.css` file, after generating the docs. Should this be ignored?

That it's what I was mentioning this morning.

The code of our CSS files would need to be updated in order to meet the linting specifications we're trying to enforce.

I propose to leave this action for a future task, that could include the implementation of a linting step on every push (?).

**Comment by xripoll-at-wiris** [comment]
> I propose to leave this action for a future task

Agree, let's not get out of scope.

**Review by xripoll-at-wiris — APPROVED** [review]

**Comment by carla-at-wiris** [comment]
> > I guess this is expected, but the `npm run lint` command in the root fails with the `packages/mathtype-html-integration-devkit/out/styles/main.css` file, after generating the docs. Should this be ignored?
> 
> That it's what I was mentioning this morning.
> 
> The code of our CSS files would need to be updated in order to meet the linting specifications we're trying to enforce.
> 
> I propose to leave this action for a future task, that could include the implementation of a linting step on every push (?).

The CSS code here shouldn't really be linted for us, in my opinion, since it comes from the documentation

File: .github/workflows/jsdoc-Generate-documentation-site.yml (hunk 1 above)
Comment (dcanet-at-wiris) [reply .github/workflows/jsdoc-Generate-documentation-site.yml:64]: No, I didn't

Comment (dcanet-at-wiris) [reply .github/workflows/jsdoc-Generate-documentation-site.yml:64]: From the documentation, it seems that the element can have any value.
https://github.com/actions/toolkit/tree/master/packages/core

**Pull request merged**